keps/sig-instrumentation/3077-contextual-logging/README.md

```diff
@@ -0,0 +1,955 @@
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Contextual logging replaces the global logger instance in klog with a
+`logr.Logger` instance that gets passed from a function’s caller into the
+function. Contextual logging builds on top of [structured
+logging](https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/1602-structured-logging).
+
+It enables the caller to:
+- attach key/value pairs that get included in all log messages
+- add names that describe which component or operation triggered a log messages
+- reduce the amount of log messages emitted by a callee by changing the
+  verbosity
+
+This works without having to pass any additional information into the callees
+because the additional information or settings are stored in the modified
+logger instance that is passed to it.
+
+Third-party components that use Kubernetes packages like client-go are no
+longer forced to use klog. They can choose an arbitrary implementation of
+`logr.Logger` and configure it as desired.
+
+During unit testing, each test case can use its own logger to ensure that log
+output is associated with the test case.
+
+## Motivation
+
+### Goals
+
+- Remove `k8s.io/klog` imports and usage from all packages.
+- Grant the caller of a function control over logging inside that function.
+- Provide documentation and helper code for setting up logging in unit tests.
+- Change as few exported APIs as possible.
+
+### Non-Goals
+
+- Remove the klog text output format.
+
+## Proposal
+
+Log calls that were already converted to structured logging need to be updated
+to use a logger that gets passed into the function in one of two ways:
+- as explicit parameter
+- attached to a `context.Context`
+
+When a function already accepts a context parameter, then that will be used
+instead of adding a separate parameter. This covers most of client-go and
+avoids another big breaking change for the community.
+
+When a function does not accept a context parameter but needs a context for
+calling some other functions, then a context parameter will be added. As a
+positive side effect, such a change can then also remove several `context.TODO`
+calls (currently over 6000 under `pkg`, `cmd`, `staging` and `test`). An
+explicit logger parameter is suitable for functions which don’t need a context
+and never will.
+
+The rationale for not using both context and an explicit logger parameter is
```

Comment:
We could do that if we want to enforce it. It should fit into the existing `logcheck` tool.

My concern is that (as with any linter) that perhaps there might be a legitimate reason why in some part of the code both a logger and a context need to be passed. Such a tool would then prevent that unless we also add a way to suppress the check per function.

Reply:
I've added a comment saying that logcheck will check for this.

I have the check ready and also integrated logcheck with golangci-lint, which adresses my concern about valid usages of the pattern because those then can use `nolint:logcheck`.